codes/rust/chapter_tree/binary_search_tree.rs

```diff
@@ -0,0 +1,201 @@
+/**
+ * File: binary_search_tree.rs
+ * Created Time: 2023-04-20
+ * Author: xBLACKICEx (xBLACKICE@outlook.com)
+ */
+use std::{cell::RefCell, rc::Rc};
+use tree_node::TreeNode;
+include!("../include/include.rs");
+type TreeNodeRc = Rc<RefCell<TreeNode>>;
+
+pub struct BinarySearchTree {
+    root: Option<TreeNodeRc>,
+}
+
+impl BinarySearchTree {
+    pub fn new(mut num: Vec<i32>) -> Self {
+        num.sort();                                 // 排序数组
+        if num.is_empty() {
+            Self { root: None }
+        } else {
+            Self {
+                root: Some(Self::build_tree(&num)), // 构建二叉搜索树
+            }
+        }
+    }
+
+    /* 获取二叉树根节点 */
+    pub fn get_root(&self) -> Option<TreeNodeRc> {
+        self.root.clone() // RC 克隆
+    }
+
+    /* 构建二叉搜索树 */
+    fn build_tree(num: &[i32]) -> TreeNodeRc {
+        // 将数组中间节点作为根节点
+        let mid = num.len() / 2;
+        let root = TreeNode::new(num[mid]);
+        // 递归建立左子树和右子树
+        if mid > 0 {
+            root.borrow_mut().left = Some(Self::build_tree(&num[..mid]));
+        }
+        if mid < num.len() - 1 {
+            root.borrow_mut().right = Some(Self::build_tree(&num[mid + 1..]));
+        }  
+        root
+    }
+
+    /* 查找节点 */
+    pub fn search(&self, num: i32) -> Option<TreeNodeRc> {
+        let mut cur = self.root.clone();
+
+        // 循环查找，越过叶节点后跳出
+        while let Some(node) = cur.clone() {
+            // 目标节点在 cur 的右子树中
+            if node.borrow().val < num {
+                cur = node.borrow().right.clone();
+            }
+            // 目标节点在 cur 的左子树中
+            else if node.borrow().val > num {
+                cur = node.borrow().left.clone();
+            }
+            // 找到目标节点，跳出循环
+            else {
+                break;
+            }
+        }
+        // 返回目标节点
+        cur
+    }
+
+    /* 插入节点 */
+    pub fn insert(&mut self, num: i32) -> Option<TreeNodeRc> {
+        // 若树为空，直接提前返回
+        if self.root.is_none() {
+            return None;
+        }
+
+        let mut cur = self.root.clone();
```

Comment:
I'm confused about `clone()` . Is it clone a new object of the `root` node? Is it necessary?